### PR TITLE
go1.8 compliant functions in vitessdriver

### PR DESCRIPTION
**Pull request by sougou**
I've left out the row support functions because most of them are
not supported by the protocol. The most important changes
are supported: contexts and named arguments.

* Deleted the deprecated shard specific connection support.
* Built go1.8, fixed compilation errors and added tests.
* Fixed documentation that was still stating support for v1 API.
* Fixed custom_sharding test that was relying on v1 API.
* Doc comments updated to explain isolation levels and named args.
* Code review comments addressed.

### CODE REVIEW

**Comment by sougou** [comment]
@kardianos

File: go/vt/vitessdriver/driver_go18.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+		return nil, err
+	}
+	qr, err := c.exec(ctx, query, bv)
+	if err != nil {
+		return nil, err
+	}
+	return result{int64(qr.InsertID), int64(qr.RowsAffected)}, nil
+}
+
+func (c *conn) QueryContext(ctx context.Context, query string, args []driver.Value) (driver.Rows, error) {
```
Comment (kardianos) [inline go/vt/vitessdriver/driver_go18.go]: I think args should be []driver.NamedValue.

File: go/vt/vitessdriver/driver_go18.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+	}
+
+	qr, err := c.exec(ctx, query, bv)
+	if err != nil {
+		return nil, err
+	}
+	return newRows(qr), nil
+}
+
+func (s *stmt) ExecContext(ctx context.Context, args []driver.Value) (driver.Result, error) {
```
Comment (kardianos) [inline go/vt/vitessdriver/driver_go18.go]: args should be []driver.NamedValue

File: go/vt/vitessdriver/driver_go18.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+	return s.c.ExecContext(ctx, s.query, args)
+}
+
+func (s *stmt) QueryContext(ctx context.Context, args []driver.Value) (driver.Rows, error) {
```
Comment (kardianos) [inline go/vt/vitessdriver/driver_go18.go]: args []driver.NamedValue

File: go/vt/vitessdriver/driver_go18.go (hunk 1 above)
Comment (kardianos) [inline go/vt/vitessdriver/driver_go18.go]: you can also include:
```
var _ driver.QueryerContext = &conn{} // Type-check interface.
```

File: go/vt/vitessdriver/driver_go18.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+import (
+	"context"
+	"errors"
+	"fmt"
+)
+
+const noIntermixingErr = errors.New("named and positional arguments intermixing disallowed")
+
+func (c *conn) ExecContext(ctx context.Context, query string, args []driver.Value) (driver.Result, error) {
```
Comment (kardianos) [inline go/vt/vitessdriver/driver_go18.go]: args should be []driver.NamedValue

File: go/vt/vitessdriver/driver_go18.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+			// Verify that there's no intermixing.
+			if nameUsed && args.Name == "" {
+				return nil, noIntermixingErr
+			}
+			if !nameUsed && args.Name != "" {
+				return nil, noIntermixingErr
+			}
+		}
+		if args.Name == "" {
+			bv[fmt.Sprintf("v%d", i+1)] = v
```
Comment (kardianos) [inline go/vt/vitessdriver/driver_go18.go]: bv[strconv.Itoa(v.Ordinal)] = v

File: go/vt/vitessdriver/driver_go18.go [hunk 6]
```diff
@@ -0,0 +1,94 @@
+// Copyright 2016, Google Inc. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// +build go1.8
+
+// TODO(sougou): Merge this with driver.go once go1.7 is deprecated.
```
Comment (kardianos) [inline go/vt/vitessdriver/driver_go18.go]: Also, original non-Context methods can have an implementation `panic("See Context method")` once 1.8 is the minimum.

File: go/vt/vitessdriver/driver_go18.go [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+		} else {
+			// Verify that there's no intermixing.
+			if nameUsed && args.Name == "" {
+				return nil, noIntermixingErr
+			}
+			if !nameUsed && args.Name != "" {
+				return nil, noIntermixingErr
+			}
+		}
+		if args.Name == "" {
```
Comment (kardianos) [inline go/vt/vitessdriver/driver_go18.go]: s/args/v/

File: go/vt/vitessdriver/driver_go18.go [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+	return s.c.QueryContext(ctx, s.query, args)
+}
+
+func bindVarsFromNamedValues(args []driver.NamedValue) (map[string]interface{}, error) {
+	bv := make(map[string]interface{}, len(args))
+	nameUsed := false
+	for i, v := range args {
+		if i == 0 {
+			// Determine if args are based on names or ordinals.
+			if args.Name != "" {
```
Comment (kardianos) [inline go/vt/vitessdriver/driver_go18.go]: s/args/v/

File: go/vt/vitessdriver/driver_go18.go [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+				return nil, noIntermixingErr
+			}
+			if !nameUsed && args.Name != "" {
+				return nil, noIntermixingErr
+			}
+		}
+		if args.Name == "" {
+			bv[fmt.Sprintf("v%d", i+1)] = v
+		} else {
+			if args.Name[0] == ":" || args.Name[0] == "@" {
```
Comment (kardianos) [inline go/vt/vitessdriver/driver_go18.go]: Probably append to user docs that no symbol prefix and if provided it will be stripped.

**Comment by alainjobart** [comment]
I thought we said we would keep the shard in there, so by specifying a specific shard, DBAs could target a shard? And if no shard is there, we revert to v3? Or do I not remember correctly? We should have written it down...

**Comment by sougou** [comment]
I don't remember either, but it's too dangerous though. They may send non-compliant statements to those shards. Since it's a public API, it could get accidentally abused.
I think it will be better if those who want to perform repairs go through vtctl.

**Comment by sougou** [comment]
@michael-berlin @alainjobart @kardianos PTAL
I've addressed all code review comments.

File: go/vt/vitessdriver/driver.go [hunk 10]
```diff
@@ -225,25 +182,29 @@ func (c *conn) dial() error {
...
+	return c.BeginContext(ctx)
+}
+
+func (c *conn) BeginContext(ctx context.Context) (driver.Tx, error) {
 	if c.Streaming {
 		return nil, errors.New("transaction not allowed for streaming connection")
 	}
-	ctx, cancel := context.WithTimeout(context.Background(), c.Timeout)
-	defer cancel()
 	tx, err := c.vtgateConn.Begin(ctx)
```
Comment (kardianos) [inline go/vt/vitessdriver/driver.go]: Does vitess support READONLY queries or IsolationLevels?

If it doesn't support those, check for them and return an error if READONLY is defined or a non-default IsolationLevel is defeined.

File: go/vt/vitessdriver/driver.go [hunk 11]
```diff
@@ -253,103 +214,105 @@ func (c *conn) Begin() (driver.Tx, error) {
 }
 
 func (c *conn) Commit() error {
+	ctx, cancel := context.WithTimeout(context.Background(), c.Timeout)
```
Comment (kardianos) [inline go/vt/vitessdriver/driver.go]: I see vitess takes a context with cancellation for Commit / Rollback(). Would it be good for the SQL driver to do the same?

File: go/vt/vitessdriver/driver.go (hunk 10 above)
Comment (sougou) [inline go/vt/vitessdriver/driver.go]: The Vitess isolation model is different from the one exposed by a traditional database. We have non-transactional reads that can be consistent or eventually consistent. Then we have traditional transactions with serializable writes. The transactional reads depend on whether the query is single-shard or not, and whether you execute them as 'FOR UPDATE'. In simpler terms, these are along the lines of what a key-value store supports.

Long story short, we don't support any of the isolation levels as defined by their contracts. So, I thought it's better to just ignore that aspect. The other option is to explicitly ensure that no isolation level was requested.

Comment (kardianos) [inline go/vt/vitessdriver/driver.go]: If there is an explicit iso set, you can define your own values in the driver for the client to use, along with a WithContext function. If it is implicit, then up to you...

The driver package requests drivers to check for the presence of both of them and return an error if not supported. But if you don't want to do that, just add a line in the doc.

**Comment by sougou** [comment]
PTAL
All new comments are addressed.

File: go/vt/vitessdriver/doc.go [hunk 12]
```diff
@@ -102,13 +54,37 @@ process of simplifying the VSchema definition and the overall process for
...
 contains more details on the VSchema:
 https://github.com/youtube/vitess/blob/master/doc/V3VindexDesign.md
+
+
+Isolation levels
+
+The Vitess isolation model is different from the one exposed by a traditional database.
+Isolation levels are controlled by connection parameters instead of Go's IsolationLevel.
+You can perform master, replica or rdonly reads. Master reads give you read-after-write
+consitency. Replica and rdonly reads give you eventual consistency. Replica reads
```
Comment (michael-berlin) [inline go/vt/vitessdriver/doc.go]: typo: consistency

File: go/vt/vitessdriver/doc.go [hunk 13]
```diff
@@ -38,57 +38,9 @@ features like consistent, application-transparent resharding:
...
-v3 automatically infers the sharding key (and therefore the target shard)
-from the query itself.
-
-It also maintains secondary indexes across shards and generates values for
-AUTO_INCREMENT columns.
+VTGate is capable of breaking up your query into parts, routing them to the
+appropriate shards and combining the results, thereby giving the semblance
+of a unified database.
 
 See the vtgate v3 Features doc for an overview:
```
Comment (michael-berlin) [inline go/vt/vitessdriver/doc.go]: I would add here a sentence which introduces the term "vtgate v3" before you're using it for the first time e.g.

> Within the Vitess codebase, this sharding transparent API is dubbed "vtgate v3".
